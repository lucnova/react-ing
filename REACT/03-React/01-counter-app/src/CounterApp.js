import React, { useState } from 'react'; // useState -> ES UN HOOK
import PropTypes from 'prop-types';

const CounterApp = ({ value = 0 }) => {
	const [counter, setCounter] = useState(value); // useState trae un arreglo con DOS VALORES [var, hacerAlgo()]

	/* ------------------------------- Handle ADD ------------------------------- */
	const handleAdd = (e) => {
		setCounter(counter + 1); // Settear un valor dado
		//setCounter((el) => el + 1 );    // Funcion callback que el retorno es seteado
	};

	/* ------------------------------- Handle SUB ------------------------------- */
	const handleSubstract = (e) => {
		setCounter(counter - 1);
	};

	/* ------------------------------ Handle Reset ------------------------------ */
	const handleReset = (e) => {
		setCounter(value);
	};

	return (
		<>
			<h1>CounterApp</h1>
			<h2>{counter}</h2>

			<button onClick={handleAdd}>+1</button>
			<button onClick={handleSubstract}>-1</button>
			<button onClick={handleReset}>R</button>
		</>
	);
};

CounterApp.propTypes = {
	value: PropTypes.number,
};

export default CounterApp;
