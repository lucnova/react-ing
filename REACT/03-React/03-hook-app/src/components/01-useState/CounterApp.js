import React, { useState } from 'react';
import './CounterApp.css';

export const CounterApp = () => {
    // HOOK: useState => Define un estado de una variable.
	const [state, setState] = useState({
		firstCounter: 10,
		secondCounter: 20,
        thirdCounter: 30,
	});

    // Como solo harÃ© uso de estas variables, las extraigo
	const { firstCounter, secondCounter } = state;

	return (
		<>
			<div className="row">
				<div className="col text-center">
					<h1>First Counter {firstCounter}</h1>
					<h1>Second Counter {secondCounter}</h1>
					<hr />

					<button
						className="btn btn-primary"
						onClick={() => {
							/* SetState -> Reemplaza el estado anterior por el nuevo estado */
							const newCounter = {
								...state,
								firstCounter: firstCounter + 1,
							};
							setState(newCounter);
						}}
					>
						+1
					</button>
				</div>
			</div>
		</>
	);
};
