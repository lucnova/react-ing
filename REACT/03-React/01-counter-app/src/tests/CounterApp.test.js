/**
 * PRUEBAS SOBRE COUNTERAPP
 *      - Probar que CounterApp sea mostrado correctamente (tal y como el snapshot)
 *
 *      - Debe mostrar el valor enviado tal cual
 */

import { shallow } from 'enzyme';
import CounterApp from '../CounterApp';

describe('* TESTS ON <CounterApp />', () => {
	let wrapper = shallow(<CounterApp />); // Para no perder intellisense si se declara undefined

	beforeEach(() => {
		wrapper = shallow(<CounterApp />);
	});

	test('should display <CounterApp /> as intended', () => {
		expect(wrapper).toMatchSnapshot();
	});

	test('should display sent initial value', () => {
		const sentValue = 500;
		const customWrapper = shallow(<CounterApp value={sentValue} />);

		const foundValue = customWrapper.find('h2').text().trim();

		expect(foundValue).toBe(sentValue.toString());
	});

	/* --------------------------- SE EJECUTA EN ORDEN -------------------------- */

	test('should increment value by 1', () => {
		wrapper.find('button').first().simulate('click');

		const foundValue = wrapper.find('h2').text().trim();
		expect(foundValue).toBe('1');
	});

	test('should decrement value by -1', () => {
		wrapper.find('button').at(1).simulate('click');

		const foundValue = wrapper.find('h2').text().trim();
		expect(foundValue).toBe('-1');
	});

	test('should reset value to the one being sended', () => {
		const sentValue = 50; // Definir el valor por defecto
		const customWrapper = shallow(<CounterApp value={sentValue} />); // Definir el componente

		// Simular varios +1
		customWrapper.find('button').first().simulate('click');
		customWrapper.find('button').first().simulate('click');
		customWrapper.find('button').first().simulate('click');
        // Simular -1
        customWrapper.find('button').at(1).simulate('click');

        // Simular reset
		customWrapper.find('button').last().simulate('click');

        // Buscar el valor y verificar que es el default
		const foundValue = customWrapper.find('h2').text().trim();
		expect(foundValue).toBe(sentValue.toString());
	});
});
