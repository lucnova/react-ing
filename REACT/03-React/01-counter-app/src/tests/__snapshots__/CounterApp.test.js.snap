// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`* TESTS ON <CounterApp /> should display <CounterApp /> as intended 1`] = `
<Fragment>
  <h1>
    CounterApp
  </h1>
  <h2>
    0
  </h2>
  <button
    onClick={[Function]}
  >
    +1
  </button>
  <button
    onClick={[Function]}
  >
    -1
  </button>
  <button
    onClick={[Function]}
  >
    R
  </button>
</Fragment>
`;
